[cmake/ios] Packaging

fetzerch opened the pull request:
@Memphiz: This is the packaging work for IOS. This time I fear I'd need your help. I cannot get it to work because fakeroot fails to set the permissions on the package contents. Could that be an issue with El Capitan? Maybe you could try it out on your machine. Next problem is that I don't have an ios device, so I can't test the generated deb.

Other than that it's fairly similar to osx, except that we have to take care of these bundle resources. I'll see if I find a better way to handle them once we got it working.

Jenkins seems to be fine already: [job](http://jenkins.kodi.tv/job/IOS/7344/) / [deb](http://mirrors.kodi.tv/test-builds/darwin/ios/kodi-20160512-040c238-cmake_ios_packaging-ios.deb)

fetzerch committed:
[cmake/ios] Packaging

Memphiz commented:
great stuff - on my list for the weekend

Memphiz commented on project/cmake/scripts/ios/archsetup.cmake:
```diff
@@ -12,10 +12,8 @@ set(CMAKE_SYSTEM_NAME Darwin)
-  elseif(CPU STREQUAL arm64)
-    set(ARCH arm-apple-darwin)
+  if(CPU STREQUAL armv7 OR CPU STREQUAL arm64)
+    set(ARCH arm-osx)
```
arm-osx? that is nonexistant afaik ;)

fetzerch replied:
https://github.com/xbmc/xbmc/blob/master/configure.ac#L582 ;)

Memphiz replied:
this is bullshit from some addon dylibs or so ... at one point i will hopefully clean it up *grrr*

Memphiz commented:
trying to run cmake and just get:

-- Could NOT find TINYXML (missing:  TINYXML_LIBRARY TINYXML_INCLUDE_DIR) 
CMake Error at /Users/Shared/xbmc-depends/buildtools-native/share/cmake-3.3/Modules/FindPackageHandleStandardArgs.cmake:148 (message):
  Could NOT find Python (missing: PYTHON_INCLUDE_DIRS)
Call Stack (most recent call first):
  /Users/Shared/xbmc-depends/buildtools-native/share/cmake-3.3/Modules/FindPackageHandleStandardArgs.cmake:388 (_FPHSA_FAILURE_MESSAGE)
  modules/FindPython.cmake:35 (find_package_handle_standard_args)
  scripts/common/macros.cmake:153 (find_package)
  CMakeLists.txt:129 (core_require_dep)


tinyxml and python27 are built already ...

Memphiz commented:
grr line comments are not working for me anymore ... well configuring the Info.plist.in and packaging it directly doesn't work ... Info.plist must be there before xcode starts building - because xcode takes the configured Info.plist and does some other magic on it (replacing some placeholders and making a binary plist from it at the end ...)

fetzerch commented:
Could you get me the commands that you used for the build and a full log? For the info.plist I need to see what exactly needs to be done. The generated xcode project has nothing in common with our legacy xcode project.

Memphiz commented:
grep the original xcode project file for Info.plist - you should see something like:

INFOPLIST_FILE = "$(SRCROOT)/xbmc/platform/darwin/ios/Info.plist";

Thats what we need too. And before the xcode project gets built - we need to configure Info.plist.in to become Info.plist ...
Will try to do a fresh install tomorrow ... atm i just configured depends for ios as usual and run:

make -C tools/depends/target/cmakebuildsys

Memphiz committed:
[depends/tar-native] - bump to 1.2.8 - fixed "dpkg-deb: subprocess tar -cf killed by signal (Abort trap: 6)" during package on osx 10.11.x

Memphiz commented:
@fetzerch can't do a pr to your remote (isn't listed as target :( ). - please try this and see if it fixes the deb building:

https://github.com/Memphiz/xbmc/commit/10820c20d25789e8026d4955592ed95813eb322d

Memphiz commented:
@fetzerch cherry-pick this one too:
https://github.com/Memphiz/xbmc/commit/9a8f29c708e3bf34b265f5253a94a3aa943c8fd6

even if i replace the info.plist with the one from a nightly i still can't kodi to run when built with cmake. If i replace only the kodi.bin it works (so all the dependency dylibs seem fine). Still trying to figure out whats wrong buts a bit hard to tell. There are some frameworks missing in the linker line - see:

https://github.com/Memphiz/xbmc/commit/01ff6ce999bf65f948862ddb22d4d648e40dc36a (not sure if the -lresolv is supposed to go there). - i am unclear why this linked at all ImageIO is used in Texture.cpp...

Beside that it still just crashes. Without any crashreport or dump or what not. I have no idea how to figure out why it crashes. Comparing the linker line shows alot of differences though - see here:

http://pastebin.com/QYwwgRfV

Well there might have been a reason why there was never a makefile based build for ios ... if i only had at least a crashlog :(

Memphiz commented:
ok - odd thing - when i sign the app i can start it ... seems xcodebuild on jenkins does something that marks the binary as "doesn't need to be signed" ... question is - what does this line do:

https://github.com/xbmc/xbmc/blob/master/tools/buildsteps/ios/make-xbmc#L11

I tried it with the jenkins deb (just signed it and installed it via xcode and it runs - this is on a jailbroken ipad though ...) - also the missing linker libraries don't seem to be an issue - odd a second time ;)

Memphiz commented:
ok found the reason - seems this line is not applied in your packaging:

https://github.com/xbmc/xbmc/blob/master/tools/darwin/Support/Codesign.command#L24

this is the fake signing taht is needed ... - beside that we only have the info plist thingy left ... not sure yet how to solve is ... this is the final info.plist we need (thats what xcode packages) ...:

http://pastebin.com/czFyST6b

Memphiz commented:
with manually applying ldid and adding the correct info.plist it looks pretty good - (just tried to watch tv via pvr addon - and it worked :) ). Really great work - only those bits missing - impressive!

fetzerch commented:
@Memphiz: Really nice that you got it working, thanks for the effort. I've rebased and added your commits. I'll try to figure out the codesign and info.plist in the next days.

fetzerch committed:
[cmake] Add missing dependencies to libcpluff

fetzerch commented:
@Memphiz: The last commit should fix the codesign

fetzerch commented:
Your commit Memphiz@10820c2 didn't fix the issue for me, I still get lots of errors that seem fakeroot related.

```
+ /Users/Shared/xbmc-depends/buildtools-native/../buildtools-native/bin/fakeroot chown -R 0:0 ./org.xbmc.kodi-ios
chown: ./org.xbmc.kodi-ios/Applications/Kodi.app/AppData/AppHome/addons/audioencoder.xbmc.builtin.aac/addon.xml: Operation not permitted
chown: ./org.xbmc.kodi-ios/Applications/Kodi.app/AppData/AppHome/addons/audioencoder.xbmc.builtin.aac/icon.png: Operation not permitted
chown: ./org.xbmc.kodi-ios/Applications/Kodi.app/AppData/AppHome/addons/audioencoder.xbmc.builtin.aac/resources/language/English/strings.po: Operation not permitted
chown: ./org.xbmc.kodi-ios/Applications/Kodi.app/AppData/AppHome/addons/audioencoder.xbmc.builtin.aac/resources/language/English: Operation not permitted
chown: ./org.xbmc.kodi-ios/Applications/Kodi.app/AppData/AppHome/addons/audioencoder.xbmc.builtin.aac/resources/language: Operation not permitted
chown: ./org.xbmc.kodi-ios/Applications/Kodi.app/AppData/AppHome/addons/audioencoder.xbmc.builtin.aac/resources/settings.xml: Operation not permitted
```

Memphiz commented:
odd i didn't see thos chown errors on my system - what is the build path you are using? (mine is /Users/username/src/kodi-cmake)

fetzerch commented:
For me it's /Users/username/develop/kodi/kodi-build-ios. I'll see if I find out why fakeroot fails, but then it's good to know that it's not necessarily an el capitan problem.

Memphiz commented:
@fertzech maybe it has something to do with the rootleds feature - http://apple.stackexchange.com/questions/193368/what-is-the-rootless-feature-in-el-capitan-really

though that folder shouldn't be under protection - i am not sure - maybe i have turned this off on my hackbook...

Memphiz commented:
@fetzerch confirmed - it has to do with the rootless / csr feature of el capitan. I had csr disabled (its disabled automagically on hackintoshs) - thats why i didn't see the problem - enabled it now and can reproduce it. I will see if we can circumvent it - but i doubt it (as its goal is to prevent unwanted root actions).

Memphiz committed:
- only use fakeroot if SIP (System Intgrity Protection) is disabled (it prevents fakeroot et al - fall back to sudo in that case) - also remove the libfauxsu approach as it has the same SIP problem and is unused since fakeroot was added

Memphiz commented:
@fetzerch here you go:

https://github.com/Memphiz/xbmc/commit/db5a74d73a449a81f1ff1c796af23b22993c6473

This detects SIP and falls back to sudo (asking for a password obviously). Need to think of that one once i ever want to upgrade to el capitan on the jenkins slaves.

Memphiz committed:
- make Info.plist.in compatible with cmake by not relying on any xcode string replacements

Memphiz commented on project/cmake/scripts/ios/install.cmake:
```diff
@@ -61,10 +61,14 @@ configure_file(${CORE_SOURCE_DIR}/tools/darwin/packaging/migrate_to_kodi_ios.sh.
-    #                                           ${CMAKE_BINARY_DIR}/tools/darwin/packaging/media/osx/
+    COMMAND "XBMC_DEPENDS_ROOT=${NATIVEPREFIX}/.."
+            "PLATFORM_NAME=${PLATFORM}"
+            "CODESIGNING_FOLDER_PATH=org.xbmc.kodi-ios/Applications/${APP_NAME}.app"
```
that can't work i think ... will try to figure out the correct path ...

fetzerch commented:
Thanks much, that indeed fixes the problem. Cherry-picked the change here.

Memphiz commented:
@fetzerch top 2 commits from here and we should be fine:

https://github.com/Memphiz/xbmc/commits/cmake_ios_packaging

please kick a jenkins build and let me verify one more time then ...

fetzerch commented:
Nice! You're the man. I was just about to look into and learn about the Info.plist stuff.

with cmake: http://jenkins.kodi.tv/job/IOS/7388/
without cmake to test for regressions: http://jenkins.kodi.tv/job/IOS/7389/

fetzerch commented:
Normal build failed because of the tar-native commit: http://jenkins.kodi.tv/job/IOS/7389/consoleFull

```
checking sys/modem.h usability... Can't exec "/Users/Shared/jenkins/workspace/IOS/tools/depends/xbmc-depends/buildtools-native/bin/aclocal": No such file or directory at /Users/Shared/jenkins/workspace/IOS/tools/depends/xbmc-depends/buildtools-native/share/autoconf/Autom4te/FileUtils.pm line 326.
autoreconf: failed to run /Users/Shared/jenkins/workspace/IOS/tools/depends/xbmc-depends/buildtools-native/bin/aclocal: No such file or directory
config.status: creating tests/automake-1.11
make[2]: *** [x86_64-osx-native] Error 1
```

Can that be something specific to the OSX-i5-1-macminivault slave? Because the other build I triggered seemed to be able to build depends.

Memphiz commented:
trying it again on that slave - if aclocal was not there this would mean automake-native didn't build ... strange thing - shouldn't be related to tar imo ... aclocal is there normally ...

Memphiz commented:
looks good to me - runtime tested aswell :) squash and go i'd say

fetzerch commented:
Squashed the ones that belong together and rebased. Thanks again much for your help @Memphiz!
